[SPARK-22181][SQL]Adds ReplaceExceptWithFilter rule

sathiyapk opened the pull request:
## What changes were proposed in this pull request?

Adds a new optimisation rule 'ReplaceExceptWithNotFilter' that replaces Except logical with Filter operator and schedule it before applying 'ReplaceExceptWithAntiJoin' rule. This way we can avoid expensive join operation if one or both of the datasets of the Except operation are fully derived out of Filters from a same parent.

## How was this patch tested?

The patch is tested locally using spark-shell + unit test.

sathiyapk committed:
SPARK-22181 Adds ReplaceExceptWithNotFilter rule

AmplabJenkins commented:
Can one of the admins verify this patch?

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1243,53 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+ * flipping the filter condition of the right child.
+ * {{{
+ *   SELECT a1, a2 FROM Tab1 WHERE a2 = 12 EXCEPT SELECT a1, a2 FROM Tab1 WHERE a1 = 5
+ *   ==>  SELECT a1, a2 FROM Tab1 WHERE a2 = 12 AND a1 <> 5
```
This is wrong. See the correct description of `EXCEPT`

> EXCEPT returns distinct rows from the left input query that aren’t output by the right input query.

sathiyapk replied:
@gatorsmile Thanks for the comment. Sorry i didn't pay attention, i'll update the query example. But the code produces the correct behaviour of the Except operator. Do you have any comment in the code?

gatorsmile replied:
I can see the value of this rule. 

cc @gengliangwang Review this?

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #82590 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82590/testReport)** for PR 19451 at commit [`1baecfd`](https://github.com/apache/spark/commit/1baecfdce9552b1e7853ace425970ea49c8f2304).

SparkQA commented:
**[Test build #82590 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82590/testReport)** for PR 19451 at commit [`1baecfd`](https://github.com/apache/spark/commit/1baecfdce9552b1e7853ace425970ea49c8f2304).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82590/
Test FAILed.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1243,53 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+    case x => x
+  }
+
+  def combineFilters(plan: LogicalPlan): LogicalPlan = CombineFilters(plan) match {
```
I think `CombineFilters(plan)` can finish in one pass.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1243,53 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+
+  def replaceAttributesIn(condition: Expression, leftChild: LogicalPlan): Expression = {
+    condition transform {
+      case AttributeReference(name, _, _, _) =>
```
See https://github.com/databricks/scala-style-guide#pattern-matching

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1243,53 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+  def replaceAttributesIn(condition: Expression, leftChild: LogicalPlan): Expression = {
+    condition transform {
+      case AttributeReference(name, _, _, _) =>
+        leftChild.output.find(_.name == name).get
```
The time complexity here is O(n^2). I suggest that we can have a hashmap from name to AttributeReference for the leftChild.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1243,53 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+    case result => result
+  }
+
+  def replaceAttributesIn(condition: Expression, leftChild: LogicalPlan): Expression = {
```
find a better param name for `leftChild`

sathiyapk committed:
SPARK-22181 Addresses Review comments

SparkQA commented:
**[Test build #82605 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82605/testReport)** for PR 19451 at commit [`ea7b150`](https://github.com/apache/spark/commit/ea7b15061bbb8dadb59d55084e6399c492194d03).

sathiyapk commented:
@gengliangwang thanks for your review. I pushed a new commit, please let me know if it addresses your comments. And regarding

> Also please use semanticEquals instead of sameResult

I couldn't call `semanticEquals ` directly on a LogicalPlan, can you please give an example how to use it efficiently?

> I think CombineFilters(plan) can finish in one pass.

CombineFilters(plan) combines two neighbouring filters into one, so assuming there can be any number of filters, we would need a recursive call, so more than one pass. Correct me if i'm wrong, thanks.

SparkQA commented:
**[Test build #82605 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82605/testReport)** for PR 19451 at commit [`ea7b150`](https://github.com/apache/spark/commit/ea7b15061bbb8dadb59d55084e6399c492194d03).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82605/
Test FAILed.

sathiyapk committed:
SPARK-22181 adds import scala.language.implicitConversions

SparkQA commented:
**[Test build #82606 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82606/testReport)** for PR 19451 at commit [`0f9ea7c`](https://github.com/apache/spark/commit/0f9ea7c6f21fd2c2e19dab73e0bec29652898757).

gengliangwang commented:
@sathiyapk Yes you're right, I have some misunderstandings. 
Still, I have other comments.

SparkQA commented:
**[Test build #82606 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82606/testReport)** for PR 19451 at commit [`0f9ea7c`](https://github.com/apache/spark/commit/0f9ea7c6f21fd2c2e19dab73e0bec29652898757).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82606/
Test PASSed.

sathiyapk committed:
SPARK-22181 Addresses Review comments

SparkQA commented:
**[Test build #82645 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82645/testReport)** for PR 19451 at commit [`5ca1e88`](https://github.com/apache/spark/commit/5ca1e882beb1b23af55a20257a1a08ec6a55de8e).

sathiyapk commented:
@gengliangwang Ready for a next review :)

> put case ... in a new line

Are your sure? I thought according to the coding style, while calling on a partial function if there is only one case, we put the case on the same line as the function invocation. isn't it?

PS: `nonFilterChild` is really a perfect name, thanks.

SparkQA commented:
**[Test build #82645 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82645/testReport)** for PR 19451 at commit [`5ca1e88`](https://github.com/apache/spark/commit/5ca1e882beb1b23af55a20257a1a08ec6a55de8e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1243,57 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+  import scala.annotation.tailrec
+  import scala.language.implicitConversions
+
+  implicit def nodeToFilter(node: LogicalPlan): Filter = node.asInstanceOf[Filter]
```
avoid using implicit
https://github.com/databricks/scala-style-guide#implicits

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1243,57 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+ */
+object ReplaceExceptWithFilter extends Rule[LogicalPlan] {
+
+  import scala.annotation.tailrec
```
maybe put this import to the beginning of file?

gengliangwang commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceOperatorSuite.scala:
```diff
@@ -50,6 +52,25 @@ class ReplaceOperatorSuite extends PlanTest {
     comparePlans(optimized, correctAnswer)
   }
 
+  test("replace Except with Filter") {
```
consider add a case that left child is not filter

gengliangwang commented:
@sathiyapk You are right about the style, I didn't notice that. 
I just left some more comments.  I think this is good to me after you fix them.

sathiyapk committed:
SPARK-22181 Addresses Review comments

SparkQA commented:
**[Test build #82703 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82703/testReport)** for PR 19451 at commit [`24b6b6c`](https://github.com/apache/spark/commit/24b6b6c519fa87f5d33d920eed3c1d43c74004a8).

gengliangwang commented:
LGTM.
@gatorsmile please help check again.

sathiyapk commented:
@gengliangwang thanks!
I'll squash my commits into 2 if everything is fine..

SparkQA commented:
**[Test build #82703 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82703/testReport)** for PR 19451 at commit [`24b6b6c`](https://github.com/apache/spark/commit/24b6b6c519fa87f5d33d920eed3c1d43c74004a8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82703/
Test PASSed.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1244,53 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+      )
+  }
+
+  def isEligible(left: LogicalPlan, right: LogicalPlan): Boolean = (left, right) match {
```
please add `private` in all the internal functions?

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1244,53 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+    case _ => false
+  }
+
+  def nonFilterChild(plan: LogicalPlan): LogicalPlan = plan.find(!_.isInstanceOf[Filter]).get
```
Nit: no need to add the return type `: LogicalPlan` for the private functions.

gatorsmile replied:
Although it is impossible, but please use
```Scala
      plan.find(!_.isInstanceOf[Filter]).getOrElse {
        throw new AnalysisException("xyz")
      } 
```

rxin replied:
it shouldn't be an analysisexception, if there is a bug in catalyst.

rxin replied:
but i agree we should throw a more meaningful error message in case there is a bug

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1244,53 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+
+  def combineFilters(plan: LogicalPlan): LogicalPlan = {
+    @tailrec
+    def fixedPoint(plan: LogicalPlan, acc: LogicalPlan): LogicalPlan = {
```
The func name is confusing...

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1244,53 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+    fixedPoint(plan, CombineFilters(plan))
+  }
+
+  def replaceAttributesIn(condition: Expression, node: LogicalPlan): Expression = {
```
We do not need this function. Could we inline these logics in the rule?

gatorsmile reviewed (commented):
Could you please add an end-to-end testsuite `except.sql` of `SQLQueryTestSuite.scala`? 

https://github.com/apache/spark/blob/720c94fe774431a8a40215757ded3dae9f267c7f/sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala

rxin commented:
Can you update the pr description with an example plan before / after this optimization, and also put that example in the comment section of the doc.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1244,53 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+ * Note:
+ * 1. We should combine all the [[Filter]] of the right node before flipping it using NOT operator.
+ */
+object ReplaceExceptWithFilter extends Rule[LogicalPlan] {
```
I'd put this in a new file. the optimizer file is getting too long.

rxin commented:
Actually you already have it in the classdoc, so please just update the pr description with it.

sathiyapk committed:
SPARK-22181 Addresses Review comments

SparkQA commented:
**[Test build #82740 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82740/testReport)** for PR 19451 at commit [`5facb93`](https://github.com/apache/spark/commit/5facb93d197cd3aeddb76813381a4499c8545a99).

sathiyapk commented:
@gatorsmile 
> Could you please add an end-to-end testsuite except.sql of SQLQueryTestSuite.scala?

Please verify `except.sql ` and  `except.sql.out` files are enough for the end-to-end testsuite.

sathiyapk commented:
@rxin I think it would be better to keep all the rules of the "Replace Operators" batch in a single file. So if you prefer to keep the rule in a new file, we can move all the replace operator rules in a new file say., `ReplaceOperators` under `*catalyst.optimizer` ?

rxin commented:
If we have to do this all over again i'd put all rules in their own files. Replace isn't really a great high level category because all rules at some level replace something.

SparkQA commented:
**[Test build #82740 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82740/testReport)** for PR 19451 at commit [`5facb93`](https://github.com/apache/spark/commit/5facb93d197cd3aeddb76813381a4499c8545a99).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82740/
Test PASSed.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1244,51 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+ * Note:
+ * 1. We should combine all the [[Filter]] of the right node before flipping it using NOT operator.
+ */
+object ReplaceExceptWithFilter extends Rule[LogicalPlan] {
```
Please move it to a separate file.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1242,6 +1244,51 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+  }
+
+  private def nonFilterChild(plan: LogicalPlan) = plan.find(!_.isInstanceOf[Filter]).getOrElse {
+    throw new IllegalStateException("LogicalPlan with no Local or Logical Relation")
```
-> `throw new AnalysisException("Leaf node is expected.")`

rxin replied:
i left a comment before - we shouldn't throw analysis exception if it an internal bug

gatorsmile commented on sql/core/src/test/resources/sql-tests/inputs/except.sql:
```diff
@@ -0,0 +1,43 @@
+
+SELECT * FROM t4;
+
+SELECT * FROM t5;
```
Line 14 to 28 are not needed. You can embed the temp view into the line 31 -43 queries.

gatorsmile commented on sql/core/src/test/resources/sql-tests/inputs/except.sql:
```diff
@@ -0,0 +1,43 @@
+
+
+-- Except operation that will be replace by left anti join
+SELECT * FROM t5 EXCEPT SELECT * FROM t1;
```
Add an empty line.

gatorsmile commented on sql/core/src/test/resources/sql-tests/inputs/except.sql:
```diff
@@ -0,0 +1,43 @@
+create temporary view t2 as select * from values
+  ("one", 1),
+  ("two", 22),
+  ("one", 5)
```
Also add a case for `NULL`.  Thanks!

sathiyapk commented:
@gatorsmile one quick question, could you please say which rule is transforming the filter conditions like `'b === 2` to `'b.isNotNull && 'b === 2` ? thanks!

gengliangwang commented:
Hi @sathiyapk ,  it is from `InferFiltersFromConstraints`.

sathiyapk commented:
@gengliangwang thank you very much!

sathiyapk committed:
SPARK-22181 Addresses Review comments

SparkQA commented:
**[Test build #82812 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82812/testReport)** for PR 19451 at commit [`95a7d46`](https://github.com/apache/spark/commit/95a7d4610bff81faf45696d88fe1c066dfddd6bd).

sathiyapk commented:
In the previous version of the rule, the `InferFiltersFromConstraints` rule added some additional filters (like 'col.isNotNull) after the filter condition is flipped, so it filtered out the null values. By applying the `InferFiltersFromConstraints` rule before flipping the condition fixes it. Please let me know if you have any questions. Thanks to @gatorsmile for reminding to add `NULL` values in the test suite.

sathiyapk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  private def nonFilterChild(plan: LogicalPlan) = plan.find(!_.isInstanceOf[Filter]).getOrElse {
+    throw new IllegalStateException("Leaf node is expected")
```
Following @rxin comment, i'm leaving it as `IllegalStateException `, please let me know if you have any other suggestions..

sathiyapk commented:
And could anybody please give me the needed rights to assign this task to myself in the jira? thanks.

gatorsmile commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceOperatorSuite.scala:
```diff
@@ -50,6 +52,42 @@ class ReplaceOperatorSuite extends PlanTest {
+
+    val correctAnswer =
+      Aggregate(table1.output, table1.output,
+        Filter(Not(attributeA >= 2 && attributeB < 1),
```
This is not a correct answer, right? We also need to consider NULL

sathiyapk committed:
SPARK-22181 Addresses Review comments

SparkQA commented:
**[Test build #82815 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82815/testReport)** for PR 19451 at commit [`1690d7d`](https://github.com/apache/spark/commit/1690d7d1509385b8bff3abce29f9ced8ca49b18a).

sathiyapk commented:
@gatorsmile you are right! actually i forgot to push the updates in the test suite. I pushed it now and the test cases are passing..

SparkQA commented:
**[Test build #82812 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82812/testReport)** for PR 19451 at commit [`95a7d46`](https://github.com/apache/spark/commit/95a7d4610bff81faf45696d88fe1c066dfddd6bd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82812/
Test FAILed.

SparkQA commented:
**[Test build #82815 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82815/testReport)** for PR 19451 at commit [`1690d7d`](https://github.com/apache/spark/commit/1690d7d1509385b8bff3abce29f9ced8ca49b18a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82815/
Test PASSed.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala:
```diff
@@ -0,0 +1,73 @@
+ * flipping the filter condition of the right child.
+ * {{{
+ *   SELECT a1, a2 FROM Tab1 WHERE a2 = 12 EXCEPT SELECT a1, a2 FROM Tab1 WHERE a1 = 5
+ *   ==>  SELECT DISTINCT a1, a2 FROM Tab1 WHERE a2 = 12 AND a1 <> 5
```
Could you update this?

sathiyapk committed:
SPARK-22181 Addresses Review comments

sathiyapk commented:
@gatorsmile done!

SparkQA commented:
**[Test build #82847 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82847/testReport)** for PR 19451 at commit [`9319eaa`](https://github.com/apache/spark/commit/9319eaa7c0fcf726a176ff95d9452c6297ee7fbe).

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  private def isEligible(left: LogicalPlan, right: LogicalPlan): Boolean = (left, right) match {
+    case (left, right: Filter) => nonFilterChild(left).sameResult(nonFilterChild(right))
```
This is too restrictive. For example, the children of `EXCEPT` could be `PROJECT`

> SELECT k, v FROM t1 EXCEPT SELECT k, v FROM t1 where v <> 1

SparkQA commented:
**[Test build #82847 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82847/testReport)** for PR 19451 at commit [`9319eaa`](https://github.com/apache/spark/commit/9319eaa7c0fcf726a176ff95d9452c6297ee7fbe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82847/
Test PASSed.

sathiyapk committed:
SPARK-22181 Addresses Review comments

sathiyapk commented:
@gatorsmile please let me know if you have comments regarding the new version of the rule, thanks.

SparkQA commented:
**[Test build #82887 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82887/testReport)** for PR 19451 at commit [`588e26d`](https://github.com/apache/spark/commit/588e26d989ca994156c81317b1606d766284aac5).

SparkQA commented:
**[Test build #82887 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82887/testReport)** for PR 19451 at commit [`588e26d`](https://github.com/apache/spark/commit/588e26d989ca994156c81317b1606d766284aac5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82887/
Test FAILed.

sathiyapk committed:
SPARK-22181 Addresses Review comments

SparkQA commented:
**[Test build #82962 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82962/testReport)** for PR 19451 at commit [`f6c95d3`](https://github.com/apache/spark/commit/f6c95d36c7a91937f0ec27485c8a439d518fdedd).

SparkQA commented:
**[Test build #82962 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82962/testReport)** for PR 19451 at commit [`f6c95d3`](https://github.com/apache/spark/commit/f6c95d36c7a91937f0ec27485c8a439d518fdedd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82962/
Test FAILed.

sathiyapk committed:
SPARK-22181 Addresses Review comments

SparkQA commented:
**[Test build #82964 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82964/testReport)** for PR 19451 at commit [`7030ab6`](https://github.com/apache/spark/commit/7030ab6a1ee4f798a39bc076595eb5c15c99a8ba).

SparkQA commented:
**[Test build #82964 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82964/testReport)** for PR 19451 at commit [`7030ab6`](https://github.com/apache/spark/commit/7030ab6a1ee4f798a39bc076595eb5c15c99a8ba).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82964/
Test PASSed.

sathiyapk committed:
SPARK-22181 Addresses Review comments

sathiyapk committed:
SPARK-22181 Addresses Review comments

SparkQA commented:
**[Test build #82984 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82984/testReport)** for PR 19451 at commit [`e010651`](https://github.com/apache/spark/commit/e010651721f45dd14d6a2e0f772c2493a1ec1c2e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82984/
Test PASSed.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala:
```diff
@@ -0,0 +1,114 @@
+ */
+object ReplaceExceptWithFilter extends Rule[LogicalPlan] {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
```
Please add a SQLConf `spark.sql.optimizer.replaceExceptWithFilter`. 

The logics in `apply` will be like
```Scala
    if (!plan.conf.replaceExceptWithFilter) {
      return plan
    }

    plan.transform {
     ...
```

sathiyapk replied:
You mean to add something like below in the catalyst `org.apache.spark.sql.internal.SQLConf` ?

```scala
val REPLACE_EXCEPT_WITH_FILTER = buildConf("spark.sql.optimizer.replaceExceptWithFilter")
.doc("When true, the apply function of the rule verifies, whether the right node of the except operation is of type Project followed Filter or Filter. If yes, the rule further verifies 1) Excluding the filter operations from the right (as well as the left node, if any) on the top, whether both the nodes evaluates to a same result by using the same project list if the top node is of type project, otherwise node.output. Remember a project list may contain a SubqueryExpression, so it is necessary to check it using projection 2) The filter condition don't contains any SubqueryExpression. If all the conditions are met, the rule will replace the except operation with a Filter operation by flipping the filter condition of the right node.")
.booleanConf
.createWithDefault(true)
```
and
```scala
def replaceExceptWithFilter: Boolean = getConf(REPLACE_EXCEPT_WITH_FILTER)
```

gatorsmile replied:
y

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala:
```diff
@@ -0,0 +1,114 @@
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Except(left: Project, right) if isEligible(left, right) =>
+      Project(left.projectList,
+        Distinct(Filter(Not(transformCondition(left.child, skipProject(right))), left.child)))
```
Can we remove this case? Adding `Filter` above `Project` is fine.

sathiyapk replied:
Are you sure? Because, by adding the `Filter` below the `Project`, the two adjacent filters (the flipped filter of the right node and any existing filters of the left node, if any) will be combined to one, further in the optimisation..

gatorsmile replied:
Yes. We will have the other rules to do push down and filter combination in the subsequent batches

gatorsmile replied:
You can make a try. If the filters are not pushed down and combined, we can fix the rules.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala:
```diff
@@ -0,0 +1,114 @@
+        Project(rightProjectList, nonFilterChild(skipProject(right))))
+  }
+
+  private def verifyFilterCondition(plan: LogicalPlan) = {
```
To filter out the cases of `SubqueryExpression `,  you can simply check 
```Scala
SubqueryExpression.hasCorrelatedSubquery(e)
```

@dilipbiswal Could you take a look at this rule?

sathiyapk replied:
That would be perfect, i'll do that, thanks!

sathiyapk replied:
Also, is there a way to simplify the function something like `val filterConditions = plan.takeWhile(_.isInstanceOf[Filter])` ?

dilipbiswal replied:
@gatorsmile I have a question. Is the intention to skip Filter nodes containing all kinds of subquery expression (i.e correlated or otherwise) ? If so , then we can't call the hasCorrelatedSubquery method as it only detects the correlated subquery expressions. Can we add a utility method  for this in SubqueryExpression ? We may have future usage for it.
```
def hasSubquery(e: Expression): Boolean = {
    e.find {
      case _: SubqueryExpression => true
      case _ => false
    }.isDefined
  }```

gatorsmile replied:
I am fine if we do not support any subquery expression for such a rule. 

Maybe you can review the whole rule?

dilipbiswal replied:
@gatorsmile yeah.. going through it..

dilipbiswal replied:
@sathiyapk I suspect there may be a regression in the following query. I am using the same schema as the one used in except.sql.
```
select t1.* from t1, t2 where t1.k = t2.k except select t1.* from t1, t2 where t1.k = t2.k and t1.k != 'one'
```
We are producing more rows using this new rule as we are not applying the distinct on the correct set of columns.
Also for this test case, in transformCondition function, we are encountering collision in attribute names as both side of join producing (k#,v#, k#, v#).  For this case, can we please check if logic to replace the attribute references is correct ?
@gatorsmile fyi.. please help double check.

dilipbiswal replied:
@gatorsmile @sathiyapk Do you think we  should explicitly specify the return types for all the private defs in the rule ? I just find it more readable.

sathiyapk replied:
@dilipbiswal well found! the reason for the more rows is in the case 1 of the rule, we put the Filter below the project. As @gatorsmile suggested, we should remove the first case then the Filter will be added above the project, then it will be resolved..

sathiyapk replied:
Regarding the collision in the attribute names, the transformCondition function found the right attributes in this specific example, but i agree we should test it with different corner cases. One more conservative solution to this problem could be to filter out the cases where a projectList contains SubqueryExpression like we do for the Filter condition. In any case, if the function didn't find the right attributes, spark will throw an error..

sathiyapk replied:
And I don't have any problem in explicitly specifying the return types for all the private defs if @gatorsmile is ok with that.

gatorsmile replied:
Yeah, please follow what @dilipbiswal suggested if the return type can help code readability.

sathiyapk replied:
@dilipbiswal One quick update regarding the attribute collision case, actually we didn't encountered any collision in the previous example, because we projected only t1.*. Also it looks like we will never encounter this problem..

dilipbiswal replied:
@sathiyapk Ok.. glad that its not an issue. Can you please check the content of left.output ? I remember seeing collision.  
```
left.output.map(x => (x.name, x)).toMap
```
the toMap operation was removing the collision if i remember it correctly. I just wanted to make sure we are replacing the attributes correctly.

sathiyapk replied:
@dilipbiswal You are right! I first removed the first case of the rule and then tested for collision, that's why i didn't find any collision. Without removing the first case of the rule, yes there is a collision. Although it looks sufficient now, in order to be careful about any unanticipated cases, i will also add an additional check `left.output.size == left.output.distinct.size` in the verifyCondition function.

sathiyapk committed:
SPARK-22181 Addresses Review comments

SparkQA commented:
**[Test build #83022 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83022/testReport)** for PR 19451 at commit [`ed54ffd`](https://github.com/apache/spark/commit/ed54ffd49abe0eeabfe812cc8a79823058d988bf).

sathiyapk commented:
@dilipbiswal @gatorsmile I included all your comments, please let me know if i missed anything, thanks.

SparkQA commented:
**[Test build #83022 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83022/testReport)** for PR 19451 at commit [`ed54ffd`](https://github.com/apache/spark/commit/ed54ffd49abe0eeabfe812cc8a79823058d988bf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83022/
Test PASSed.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1200,6 +1211,8 @@ class SQLConf extends Serializable with Logging {
 
   def arrowMaxRecordsPerBatch: Int = getConf(ARROW_EXECUTION_MAX_RECORDS_PER_BATCH)
 
+  def replaceExceptWithFilter: Boolean = getConf(REPLACE_EXCEPT_WITH_FILTER)
```
This is not being used, right?

sathiyapk replied:
Yes, this function is not used anywhere.

gatorsmile replied:
You need to address https://github.com/apache/spark/pull/19451/files#r146315930

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -922,6 +922,17 @@ object SQLConf {
       .createWithDefault(10000)
 
+  val REPLACE_EXCEPT_WITH_FILTER = buildConf("spark.sql.optimizer.replaceExceptWithFilter")
+    .doc("When true, the apply function of the rule verifies whether the right node of the " +
```
> `.internal()`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala:
```diff
@@ -0,0 +1,111 @@
+    left.output.size == left.output.distinct.size &&
+      verifyFilterCondition(skipProject(left)) && verifyFilterCondition(skipProject(right)) &&
+        Project(leftProjectList, nonFilterChild(skipProject(left))).sameResult(
+          Project(rightProjectList, nonFilterChild(skipProject(right))))
```
How about this?

```Scala
    left.output.size == left.output.distinct.size &&
      left.expressions.forall(!SubqueryExpression.hasSubquery(_)) &&
      right.expressions.forall(!SubqueryExpression.hasSubquery(_)) &&
      Project(projectList(right), nonFilterChild(skipProject(left))).sameResult(
        Project(projectList(right), nonFilterChild(skipProject(right))))
```

gatorsmile replied:
Then, `verifyFilterCondition ` and `collectAllExpressions ` can be removed.

sathiyapk replied:
For the rule it is enough to test SubqueryExpressions for the top Filter node(s) conditions in the tree. I think `node.expressions` will return all the expressions of the entire tree, which is actually not needed. Correct me if i'm wrong..

sathiyapk replied:
It would be nice if there is a function like `def takeWhile(p: A => Boolean)` in functional programming to substitute this logic in the verifyFilterCondition function:
```scala
    var i = 0
    val filterConditions = new collection.mutable.ArrayBuffer[Expression]
    while (plan.p(i).isInstanceOf[Filter]) {
      val condition = plan.p(i).asInstanceOf[Filter].condition
      filterConditions.insertAll(i, collectAllExpressions(condition))
      i += 1
    }
```

sathiyapk replied:
As a second thought, i think it's possible to remove the `collectAllExpressions` function by modifying the `verifyFilterCondition` function like below. But assuming there could be another filter follows, we need to iterate down, so we still need `verifyFilterCondition `.
```scala
private def verifyFilterCondition(plan: LogicalPlan): Boolean = {
    var i = 0
    val filterConditions = new collection.mutable.ArrayBuffer[Expression]
    while (plan.p(i).isInstanceOf[Filter]) {
      filterConditions.insertAll(i, plan.p(i).expressions)
      i += 1
    }
    filterConditions.forall(!SubqueryExpression.hasSubquery(_))
  }
```

sathiyapk replied:
And i didn't notice, the `left.output.size == left.output.distinct.size` in the verifyConditions function has to be changed something like 
`left.output.size == left.output.map(_.name).distinct.size`

gatorsmile replied:
I missed it. `left.expressions` only returns the expressions of the left operator. It does not include the expressions of the children.

gatorsmile replied:
Does `left.find(_.expressions.exists(SubqueryExpression.hasSubquery)).isEmpty` work?

gatorsmile replied:
The point here is to avoid the extra complexity.

sathiyapk replied:
@gatorsmile doing `left.find(_.expressions.exists(SubqueryExpression.hasSubquery)).isEmpty` will check for the `SubqueryExpression` for all the children of the left node. But it's enough to do the verification only for the successive Filter node(s).

gatorsmile replied:
Let us check all the children of the left node. This will not be an issue for 99% cases and also simplify the codes.

sathiyapk replied:
That makes sense.

gatorsmile commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelationSuite.scala:
```diff
@@ -30,6 +30,7 @@ class PropagateEmptyRelationSuite extends PlanTest {
       Batch("PropagateEmptyRelation", Once,
         CombineUnions,
         ReplaceDistinctWithAggregate,
+        ReplaceExceptWithFilter,
```
Why?

sathiyapk replied:
If you think it is not useful here, i can remove it..

sathiyapk replied:
@gatorsmile Please let me know if you think it's not useful here, also the
def replaceExceptWithFilter: Boolean = getConf(REPLACE_EXCEPT_WITH_FILTER) from the SQLConf, i'll push a new version with all the modifications, thanks.

gatorsmile replied:
This rule is not related to `PropagateEmptyRelationSuite.scala`

sathiyapk committed:
SPARK-22181 Addresses Review comments

SparkQA commented:
**[Test build #83095 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83095/testReport)** for PR 19451 at commit [`8bce960`](https://github.com/apache/spark/commit/8bce9608f8f608f3dd5d52fdffe3f5345651dd1d).

SparkQA commented:
**[Test build #83095 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83095/testReport)** for PR 19451 at commit [`8bce960`](https://github.com/apache/spark/commit/8bce9608f8f608f3dd5d52fdffe3f5345651dd1d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83095/
Test PASSed.

sathiyapk commented:
@gatorsmile I think i addressed all the comments, please let me know if i missed anything, thanks.

gatorsmile commented on sql/core/src/test/resources/sql-tests/inputs/except.sql:
```diff
@@ -0,0 +1,57 @@
+
+
+-- Except operation that will be replaced by left anti join
+SELECT * FROM t1 EXCEPT SELECT * FROM t2;
```
To see whether these SQL use the rule you added, you can add `EXPLAIN EXTENDED` before the actual query `SELECT * FROM t1 EXCEPT SELECT * FROM t2`

You will see the optimized plan and check whether the rule is applicable.

gatorsmile replied:
BTW, because our plan is not stable, I am not asking you for outputting the plan in the test cases. This is just to help you debug the codes. Thanks!

sathiyapk replied:
Thanks, good to know. I have been manually testing it via spark shell.

gatorsmile replied:
How about the result?

sathiyapk replied:
All verified via spark shell using `createOrReplaceTempView`

sathiyapk replied:
And please let me know when it is right time to squash the commits..

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala:
```diff
@@ -0,0 +1,104 @@
+  }
+
+  private def isEligible(left: LogicalPlan, right: LogicalPlan): Boolean = (left, right) match {
+    case (_, right @ (Project(_, _: Filter) | Filter(_, _))) => verifyConditions(left, right)
```
This can be further extended in the future. Let us leave a TODO mark here.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala:
```diff
@@ -0,0 +1,104 @@
+      attributeNameMap(a.name)
+    }
+
+    transformedCondition
```
```Scala
filterCondition.transform { case a: AttributeReference => attributeNameMap(a.name) }
```

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala:
```diff
@@ -0,0 +1,104 @@
+
+  private def transformCondition(left: LogicalPlan, right: LogicalPlan): Expression = {
+    val filterCondition = InferFiltersFromConstraints(combineFilters(right)
+    ).asInstanceOf[Filter].condition
```
```Scala
    val filterCondition =
      InferFiltersFromConstraints(combineFilters(right)).asInstanceOf[Filter].condition
```

gatorsmile replied:
Also need to leave a comment to explain why we need to call the rule `InferFiltersFromConstraints `

sathiyapk replied:
There is a note in the class doc regarding why do we need to call `InferFiltersFromConstraints ` rule. Comment is still needed?

sathiyapk replied:
Better like this? `* 2. Apply InferFiltersFromConstraints rule (to take into account of NULL values in the condition).`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala:
```diff
@@ -0,0 +1,104 @@
+    def iterate(plan: LogicalPlan, acc: LogicalPlan): LogicalPlan = {
+      if (acc.fastEquals(plan)) acc else iterate(acc, CombineFilters(acc))
+    }
+
```
Remove this empty line.

gatorsmile commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceOperatorSuite.scala:
```diff
@@ -50,6 +52,108 @@ class ReplaceOperatorSuite extends PlanTest {
+    val correctAnswer =
+      Aggregate(table1.output, table1.output,
+        Filter(Not((attributeA.isNotNull && attributeB.isNotNull)
+          && (attributeA >= 2 && attributeB < 1)),
```
Move `&& ` to the end of line 107

gatorsmile commented:
LGMT except a few minor comments.

gatorsmile commented:
Please change the title to `[SPARK-22181][SQL]Adds ReplaceExceptWithNotFilter rule`

sathiyapk committed:
SPARK-22181 Addresses Review comments

SparkQA commented:
**[Test build #83135 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83135/testReport)** for PR 19451 at commit [`40d8797`](https://github.com/apache/spark/commit/40d8797b1750d8a8814dcdc113fa79a3ab086cf2).

SparkQA commented:
**[Test build #83135 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83135/testReport)** for PR 19451 at commit [`40d8797`](https://github.com/apache/spark/commit/40d8797b1750d8a8814dcdc113fa79a3ab086cf2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83135/
Test PASSed.

gatorsmile commented:
Thanks for your work! 

Merged to master.

sathiyapk commented:
Thanks!
